chain-api/src/types/TokenBalance.ts

```diff
@@ -403,18 +430,107 @@ export class TokenBalance extends ChainObject {
     return { subtract };
   }
 
-  private ensureQuantityIsSpendable(quantity: BigNumber): void {
-    // locked & in use are not supported for fungibles
-    const spendableQuantity = this.quantity;
+  private ensureQuantityIsSpendable(quantity: BigNumber, currentTime: number): void {
+    // in use not supported for fungibles
+    const lockedQuantity = this.getCurrentLockedQuantity(currentTime);
+    const spendableQuantity = this.quantity.minus(lockedQuantity);
 
     if (spendableQuantity.isLessThan(quantity)) {
       throw new ValidationFailedError("Insufficient balance", {
         balanceKey: this.getCompositeKey(),
-        total: this.quantity.toFixed()
+        total: this.quantity.toFixed(),
+        lockedQuantity: lockedQuantity.toFixed()
       });
     }
   }
 
+  private ensureTokenQuantityHoldIsFungible(hold: TokenHold) {
+    if (!hold.instanceId.isEqualTo(TokenInstance.FUNGIBLE_TOKEN_INSTANCE)) {
+      const message = `Attempted to perform FT-specific operation on balance containing NFT instances`;
+      throw new ValidationFailedError(message, {
+        balanceKey: this.getCompositeKey(),
+        tokenHold: hold
+      });
+    }
+  }
+
+  public ensureCanLockQuantity(hold: TokenHold): { lock(): void } {
+    this.ensureTokenQuantityHoldIsFungible(hold);
+    this.ensureQuantityIsSpendable(hold.quantity, hold.created);
+
+    const lock = () => {
+      this.lockedHolds = [...this.getUnexpiredLockedHolds(hold.created), hold];
+    };
+
+    return { lock };
+  }
+
+  public ensureCanUnlockQuantity(
+    quantity: BigNumber,
+    currentTime: number,
+    name?: string | undefined,
+    lockAuthority?: string | undefined
+  ): { unlock(): void } {
+    const unexpiredLockedHolds = this.getUnexpiredLockedHolds(currentTime);
+
+    // sort holds in order of ascending expiration, 0 = no expiration date
+    unexpiredLockedHolds.sort((a: TokenHold, b: TokenHold) => {
+      if (b.expires === 0 && a.expires === 0) {
+        return 0;
+      } else if (b.expires === 0) {
+        return -1;
+      } else if (a.expires === 0 || a.expires > b.expires) {
+        return 1;
+      } else {
+        return -1;
+      }
+    });
```

Comment:
I wonder if we can make a separate method for that like `getUnexpiredLockedHoldsWithExpirationAscending`. I know the name is ugly, but it puts the sorting logic outside of this method and makes it easier to understand the rest

Reply:
This is a good call, I ended up moving the sort logic itself to a static method on the `TokenHold` class (it may be useful down the line for other methods / places in the code), and added `getUnexpiredLockedHoldsSortedByAscendingExpiration` as a method to get the sorted values.